boa/src/builtins/string/mod.rs

```diff
@@ -1295,34 +1327,54 @@ impl String {
     /// [mdn]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/matchAll
     /// [regex]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions
     /// [cg]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions/Groups_and_Ranges
-    // TODO: update this method to return iterator
     pub(crate) fn match_all(this: &Value, args: &[Value], context: &mut Context) -> Result<Value> {
-        let re: Value = match args.get(0) {
-            Some(arg) => {
-                if arg.is_null() {
-                    RegExp::constructor(
-                        &Value::from(Object::default()),
-                        &[Value::from(arg.to_string(context)?), Value::from("g")],
-                        context,
-                    )
-                } else if arg.is_undefined() {
-                    RegExp::constructor(
-                        &Value::from(Object::default()),
-                        &[Value::undefined(), Value::from("g")],
-                        context,
-                    )
-                } else {
-                    Ok(arg.clone())
+        // 1. Let O be ? RequireObjectCoercible(this value).
+        let object = this.require_object_coercible(context)?;
+
+        // 2. If regexp is neither undefined nor null, then
+        let regexp = args.get(0).cloned().unwrap_or_default();
+        if !regexp.is_null_or_undefined() {
+            // a. Let isRegExp be ? IsRegExp(regexp).
+            // b. If isRegExp is true, then
+            if regexp.as_object().unwrap_or_default().is_regexp() {
+                // i. Let flags be ? Get(regexp, "flags").
+                let flags = regexp.get_field("flags", context)?;
+
+                // ii. Perform ? RequireObjectCoercible(flags).
+                flags.require_object_coercible(context)?;
+
+                // iii. If ? ToString(flags) does not contain "g", throw a TypeError exception.
+                if !flags.to_string(context)?.contains('g') {
+                    return context.throw_type_error(
+                        "String.prototype.matchAll called with a non-global RegExp argument",
+                    );
                 }
             }
-            None => RegExp::constructor(
-                &Value::from(Object::default()),
-                &[Value::from(""), Value::from("g")],
-                context,
-            ),
-        }?;
-
-        RegExp::match_all(&re, this.to_string(context)?.to_string(), context)
+
+            // c. Let matcher be ? GetMethod(regexp, @@matchAll).
+            // d. If matcher is not undefined, then
+            if let Some(matcher) = regexp
+                .as_object()
+                .unwrap_or_default()
+                .get_method(context, "matchAll")?
```

Comment:
I think `"matchAll"` should instead be the `matchAll` Symbol.

Reply:
Yes, great catch. Some more tests are passing now.